### PR TITLE
Added support for UPN Suffixes to FreeRDP. 

### PR DESCRIPTION
**Pull request by Edlanta**
This includes a fix for issue: 
https://github.com/FreeRDP/FreeRDP/issues/2427

These changes are working for me. I can now logon with UPN Suffix to a win2012r2 Server and additionally to a win2012r2 Server which is behind a 2012 TS Gateway Server.
But this changes have some restrictions:
- If you want to use UPN Suffix Logon, you MUST enter a domain name with /d: even if this 
  domain name doesn't exist at all.
- If you want to use UPN Suffix Logon with TS Gateway, you MUST additionally enter a Gateway 
  domain name with /gd: even if this domain name doesn't exist at all.

This restrictions are necessary, because this is the only way we can decide if the username is for Kerberos authentication or for normal authentication with UPN suffix. I explained this in issue #2427

### CODE REVIEW

**Comment by bmiklautz** [comment]
@freerdp-bot test

File: client/common/cmdline.c [hunk 1]
```diff
@@ -1962,6 +1953,31 @@ int freerdp_client_settings_parse_command_line_arguments(rdpSettings* settings,
...
+       
+       // Check if Username is set and Domain is missing.
+       // If this is the case we should split Username and Domain at the @ sign (e.g. for Kerberos)
+       // Else we should let the username string as it is (e.g. for UPN Suffix)
+       if (username && !settings->Domain)
+       {
+	  freerdp_parse_username(username, &user, &domain);
+         
+	  settings->Username = user;
+	  settings->Domain = domain;
```
Comment (akallabeth) [inline client/common/cmdline.c]: @Edlanta got yourself a memory leak here.

File: client/common/cmdline.c [hunk 2]
```diff
@@ -1962,6 +1953,31 @@ int freerdp_client_settings_parse_command_line_arguments(rdpSettings* settings,
...
+       else
+	  settings->Username = username;
+       
+       // Same as above for Gateway Username and Domain.
+       if (gw_username && !settings->GatewayDomain)
+       {
+	  freerdp_parse_username(gw_username, &user, &domain);
+         
+	  settings->GatewayUsername = user;
+	  settings->GatewayDomain = domain;
```
Comment (akallabeth) [inline client/common/cmdline.c]: @Edlanta and here again.

**Comment by Edlanta** [comment]
Sorry for the second commit, wasnt really fit in the morning as i changed this. Now everything should work well again.

**Comment by akallabeth** [comment]
@Edlanta Ok, so you reverted the broken change, still ```freerdp_parse_username``` allocates new memory for the output parameters, but does not free up the input argument. That one is still leaking. So either only ```_strdup``` the values of ```username``` and ```gw_username``` where they are actually assigned to settings or free up the duplicated strings where they are no longer in use.

File: client/common/cmdline.c [hunk 3]
```diff
@@ -1962,6 +1953,51 @@ int freerdp_client_settings_parse_command_line_arguments(rdpSettings* settings,
...
+	  settings->GatewayUsername = _strdup(gw_username);
+       
+       if (user != NULL)
+	 free(user);
+       if (domain != NULL)
+	 free(domain);
+       if (username != NULL)
+	 free(username);
+       if (gw_username != NULL)
+	 free(gw_username);
```
Comment (nfedera) [inline client/common/cmdline.c]: You don't have to check pointers passed to free() for NULL. Please see man free ;)

**Comment by akallabeth** [comment]
Implemented this feature with #2701 in a cleaner fashion.
@Edlanta please close your pull.

**Pull request closed without merging**